[BUG] Intermittent test fail due to file writing

### Describe the bug

file writing in CI causes crashes occasionally 

https://github.com/aeon-toolkit/aeon/actions/runs/9521220064/job/26248164199?pr=1675
FAILED aeon/datasets/tests/test_tsad_data_loader.py::test_load_anomaly_detection_unavailable - OSError: Cannot save file into a non-existent directory: '/home/runner/work/aeon/aeon/aeon/datasets/local_data'

it should write to a temp directory instead, e.g. test_write uses something like this

```python
    with tempfile.TemporaryDirectory() as tmp:
        write_to_tsfile(
            X=X, path=tmp, y=y, problem_name=problem_name, regression=regression
        )
```

--- a/aeon/datasets/tests/test_tsad_data_loader.py
+++ b/aeon/datasets/tests/test_tsad_data_loader.py
@@ -1,13 +1,12 @@
 """Test anomaly detection dataset loaders."""
 
-import shutil
+import tempfile
 from pathlib import Path
 
 import numpy as np
 import pytest
 
 from aeon.datasets._tsad_data_loaders import (
-    _DATA_FOLDER,
     load_anomaly_detection,
     load_daphnet_s06r02e0,
     load_ecg_diff_count_3,
@@ -16,63 +15,77 @@ from aeon.datasets._tsad_data_loaders import (
 )
 
 
-def test_load_anomaly_detection_wrong_name():
+def test_load_anomaly_detection_wrong_name(mocker):
     """Test load non-existent anomaly detection datasets."""
     with pytest.raises(
         ValueError, match="The name of the dataset must be a tuple of two strings.*"
     ):
         load_anomaly_detection(("FOO", "BAR", "BAZ"))
-    with pytest.raises(
-        ValueError,
-        match="When loading a custom dataset, the extract_path must point "
-        "to a TimeEval-formatted CSV file.*",
-    ):
-        load_anomaly_detection(("FOO", "BAR"))
+    with tempfile.TemporaryDirectory() as tmp:
+        tmp = Path(tmp)
+        mocker.patch("aeon.datasets.tsad_datasets._DATA_FOLDER", tmp)
+        with pytest.raises(
+            ValueError,
+            match="When loading a custom dataset, the extract_path must point "
+            "to a TimeEval-formatted CSV file.*",
+        ):
+            load_anomaly_detection(("FOO", "BAR"))
 
 
-def test_load_anomaly_detection_no_train_split():
+def test_load_anomaly_detection_no_train_split(mocker):
     """Test load train split of anomaly detection dataset without training split."""
     name = ("Dodgers", "101-freeway-traffic")
-    with pytest.raises(
-        ValueError, match="Dataset .* does not have a training partition.*"
-    ):
-        load_anomaly_detection(name, split="train")
+    with tempfile.TemporaryDirectory() as tmp:
+        tmp = Path(tmp)
+        mocker.patch("aeon.datasets.tsad_datasets._DATA_FOLDER", tmp)
+        with pytest.raises(
+            ValueError, match="Dataset .* does not have a training partition.*"
+        ):
+            load_anomaly_detection(name, extract_path=tmp, split="train")
 
 
-def test_load_anomaly_detection_from_archive():
+def test_load_anomaly_detection_from_archive(mocker):
     """Test load anomaly detection dataset from archive."""
     name = ("Genesis", "genesis-anomalies")
-    X, y, meta = load_anomaly_detection(name, return_metadata=True)
-    assert isinstance(X, np.ndarray)
-    assert X.shape == (16220, 18)
-    assert isinstance(y, np.ndarray)
-    assert y.shape == (16220,)
-    assert isinstance(meta, dict)
-    assert meta["learning_type"] == "unsupervised"
-    assert meta["num_anomalies"] == 3
-    np.testing.assert_almost_equal(meta["contamination"], 0.0031, decimal=4)
+    with tempfile.TemporaryDirectory() as tmp:
+        tmp = Path(tmp)
+        mocker.patch("aeon.datasets.tsad_datasets._DATA_FOLDER", tmp)
+        X, y, meta = load_anomaly_detection(
+            name, extract_path=tmp, return_metadata=True
+        )
+        assert isinstance(X, np.ndarray)
+        assert X.shape == (16220, 18)
+        assert isinstance(y, np.ndarray)
+        assert y.shape == (16220,)
+        assert isinstance(meta, dict)
+        assert meta["learning_type"] == "unsupervised"
+        assert meta["num_anomalies"] == 3
+        np.testing.assert_almost_equal(meta["contamination"], 0.0031, decimal=4)
 
-    shutil.rmtree(_DATA_FOLDER / "multivariate" / "Genesis")
 
-
-def test_load_anomaly_detection_unavailable():
+def test_load_anomaly_detection_unavailable(mocker):
     """Test load anomaly detection dataset that is not available."""
     name = ("IOPS", "05f10d3a-239c-3bef-9bdc-a2feeb0037aa")
-    with pytest.raises(
-        ValueError, match=f"Collection {name[0]} .* not available for download.*"
-    ):
-        load_anomaly_detection(name)
+    with tempfile.TemporaryDirectory() as tmp:
+        mocker.patch("aeon.datasets.tsad_datasets._DATA_FOLDER", Path(tmp))
+        with pytest.raises(
+            ValueError, match=f"Collection {name[0]} .* not available for download.*"
+        ):
+            load_anomaly_detection(name)
 
 
-def test_load_anomaly_detection_custom_file():
+def test_load_anomaly_detection_custom_file(mocker):
     """Test load anomaly detection dataset from custom file."""
     name = ("correlation-anomalies", "rmj-2-short-2-diff-channel")
-    with pytest.raises(
-        ValueError,
-        match="When loading a custom dataset, the extract_path must point to a "
-        "TimeEval-formatted CSV file.*",
-    ):
-        load_anomaly_detection(name)
+    with tempfile.TemporaryDirectory() as tmp:
+        tmp = Path(tmp)
+        mocker.patch("aeon.datasets.tsad_datasets._DATA_FOLDER", tmp)
+        with pytest.raises(
+            ValueError,
+            match="When loading a custom dataset, the extract_path must point to a "
+            "TimeEval-formatted CSV file.*",
+        ):
+            load_anomaly_detection(name, extract_path=tmp)
 
     def _test(filename: str, learning_type: str) -> None:
         X, y, meta = load_anomaly_detection(

--- a/examples/datasets/load_data_from_web.ipynb
+++ b/examples/datasets/load_data_from_web.ipynb
@@ -31,7 +31,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 10,
+   "execution_count": 1,
    "outputs": [],
    "source": [
     "from aeon.datasets import (\n",
@@ -44,8 +44,8 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:35.937391Z",
-     "start_time": "2024-06-03T13:57:35.932399Z"
+     "end_time": "2024-06-17T13:07:18.978527Z",
+     "start_time": "2024-06-17T13:07:18.453465Z"
     }
    }
   },
@@ -68,7 +68,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 11,
+   "execution_count": 2,
    "outputs": [
     {
      "name": "stdout",
@@ -89,8 +89,8 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:35.990038Z",
-     "start_time": "2024-06-03T13:57:35.985188Z"
+     "end_time": "2024-06-17T13:07:18.985584Z",
+     "start_time": "2024-06-17T13:07:18.980660Z"
     }
    }
   },
@@ -113,7 +113,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 12,
+   "execution_count": 3,
    "outputs": [
     {
      "name": "stdout",
@@ -136,8 +136,8 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:36.020721Z",
-     "start_time": "2024-06-03T13:57:36.006084Z"
+     "end_time": "2024-06-17T13:07:19.007332Z",
+     "start_time": "2024-06-17T13:07:18.987354Z"
     }
    }
   },
@@ -184,13 +184,13 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 13,
+   "execution_count": 4,
    "outputs": [
     {
      "data": {
       "text/plain": "['AcousticContaminationMadrid',\n 'AluminiumConcentration',\n 'AppliancesEnergy',\n 'AustraliaRainfall',\n 'BIDMC32HR',\n 'BIDMC32RR',\n 'BIDMC32SpO2',\n 'BarCrawl6min',\n 'BeijingIntAirportPM25Quality',\n 'BeijingPM10Quality',\n 'BeijingPM25Quality',\n 'BenzeneConcentration',\n 'BinanceCoinSentiment',\n 'BitcoinSentiment',\n 'BoronConcentration',\n 'CalciumConcentration',\n 'CardanoSentiment',\n 'ChilledWaterPredictor',\n 'CopperConcentration',\n 'Covid19Andalusia',\n 'Covid3Month',\n 'DailyOilGasPrices',\n 'DailyTemperatureLatitude',\n 'DhakaHourlyAirQuality',\n 'ElectricMotorTemperature',\n 'ElectricityPredictor',\n 'EthereumSentiment',\n 'FloodModeling1',\n 'FloodModeling2',\n 'FloodModeling3',\n 'GasSensorArrayAcetone',\n 'GasSensorArrayEthanol',\n 'HotwaterPredictor',\n 'HouseholdPowerConsumption1',\n 'HouseholdPowerConsumption2',\n 'IEEEPPG',\n 'IronConcentration',\n 'LPGasMonitoringHomeActivity',\n 'LiveFuelMoistureContent',\n 'MadridPM10Quality',\n 'MagnesiumConcentration',\n 'ManganeseConcentration',\n 'MethaneMonitoringHomeActivity',\n 'MetroInterstateTrafficVolume',\n 'NaturalGasPricesSentiment',\n 'NewsHeadlineSentiment',\n 'NewsTitleSentiment',\n 'OccupancyDetectionLight',\n 'PPGDalia',\n 'ParkingBirmingham',\n 'PhosphorusConcentration',\n 'PotassiumConcentration',\n 'PrecipitationAndalusia',\n 'SierraNevadaMountainsSnow',\n 'SodiumConcentration',\n 'SolarRadiationAndalusia',\n 'SteamPredictor',\n 'SulphurConcentration',\n 'TetuanEnergyConsumption',\n 'VentilatorPressure',\n 'WaveDataTension',\n 'WindTurbinePower',\n 'ZincConcentration']"
      },
-     "execution_count": 13,
+     "execution_count": 4,
      "metadata": {},
      "output_type": "execute_result"
     }
@@ -203,14 +203,14 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:36.034825Z",
-     "start_time": "2024-06-03T13:57:36.029160Z"
+     "end_time": "2024-06-17T13:07:19.018724Z",
+     "start_time": "2024-06-17T13:07:19.010510Z"
     }
    }
   },
   {
    "cell_type": "code",
-   "execution_count": 14,
+   "execution_count": 5,
    "outputs": [
     {
      "name": "stdout",
@@ -227,8 +227,8 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:36.129102Z",
-     "start_time": "2024-06-03T13:57:36.049794Z"
+     "end_time": "2024-06-17T13:07:19.115945Z",
+     "start_time": "2024-06-17T13:07:19.020208Z"
     }
    }
   },
@@ -248,13 +248,13 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 15,
+   "execution_count": 6,
    "outputs": [
     {
      "data": {
       "text/plain": "['australian_electricity_demand_dataset',\n 'car_parts_dataset_with_missing_values',\n 'car_parts_dataset_without_missing_values',\n 'cif_2016_dataset',\n 'covid_deaths_dataset',\n 'covid_mobility_dataset_with_missing_values',\n 'covid_mobility_dataset_without_missing_values',\n 'dominick_dataset',\n 'elecdemand_dataset',\n 'electricity_hourly_dataset',\n 'electricity_weekly_dataset',\n 'fred_md_dataset',\n 'hospital_dataset',\n 'kaggle_web_traffic_dataset_with_missing_values',\n 'kaggle_web_traffic_dataset_without_missing_values',\n 'kaggle_web_traffic_weekly_dataset',\n 'kdd_cup_2018_dataset_with_missing_values',\n 'kdd_cup_2018_dataset_without_missing_values',\n 'london_smart_meters_dataset_with_missing_values',\n 'london_smart_meters_dataset_without_missing_values',\n 'm1_monthly_dataset',\n 'm1_quarterly_dataset',\n 'm1_yearly_dataset',\n 'm3_monthly_dataset',\n 'm3_other_dataset',\n 'm3_quarterly_dataset',\n 'm3_yearly_dataset',\n 'm4_daily_dataset',\n 'm4_hourly_dataset',\n 'm4_monthly_dataset',\n 'm4_quarterly_dataset',\n 'm4_weekly_dataset',\n 'm4_yearly_dataset',\n 'nn5_daily_dataset_with_missing_values',\n 'nn5_daily_dataset_without_missing_values',\n 'nn5_weekly_dataset',\n 'pedestrian_counts_dataset',\n 'saugeenday_dataset',\n 'solar_10_minutes_dataset',\n 'solar_4_seconds_dataset',\n 'solar_weekly_dataset',\n 'sunspot_dataset_with_missing_values',\n 'sunspot_dataset_without_missing_values',\n 'tourism_monthly_dataset',\n 'tourism_quarterly_dataset',\n 'tourism_yearly_dataset',\n 'traffic_hourly_dataset',\n 'traffic_weekly_dataset',\n 'us_births_dataset',\n 'weather_dataset',\n 'wind_4_seconds_dataset',\n 'wind_farms_minutely_dataset_with_missing_values',\n 'wind_farms_minutely_dataset_without_missing_values']"
      },
-     "execution_count": 15,
+     "execution_count": 6,
      "metadata": {},
      "output_type": "execute_result"
     }
@@ -267,14 +267,14 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:36.138465Z",
-     "start_time": "2024-06-03T13:57:36.131935Z"
+     "end_time": "2024-06-17T13:07:19.124403Z",
+     "start_time": "2024-06-17T13:07:19.117901Z"
     }
    }
   },
   {
    "cell_type": "code",
-   "execution_count": 16,
+   "execution_count": 7,
    "outputs": [
     {
      "name": "stdout",
@@ -308,8 +308,8 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:38.963800Z",
-     "start_time": "2024-06-03T13:57:36.140381Z"
+     "end_time": "2024-06-17T13:07:21.595604Z",
+     "start_time": "2024-06-17T13:07:19.126358Z"
     }
    }
   },
@@ -345,17 +345,17 @@
     "from aeon.datasets.tsad_datasets import multivariate, univariate\n",
     "\n",
     "# This file also contains sub lists by learning type, e.g. semi-supervised, ...\n",
-    "print(\"Univariate length = \", len(univariate))\n",
-    "print(\"Multivariate length = \", len(multivariate))"
+    "print(\"Univariate length = \", len(univariate()))\n",
+    "print(\"Multivariate length = \", len(multivariate()))"
    ],
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:38.971399Z",
-     "start_time": "2024-06-03T13:57:38.966037Z"
+     "end_time": "2024-06-17T13:07:21.771910Z",
+     "start_time": "2024-06-17T13:07:21.597146Z"
     }
    },
-   "execution_count": 17
+   "execution_count": 8
   },
   {
    "cell_type": "markdown",
@@ -397,11 +397,11 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:39.104489Z",
-     "start_time": "2024-06-03T13:57:38.973303Z"
+     "end_time": "2024-06-17T13:07:22.224850Z",
+     "start_time": "2024-06-17T13:07:21.773453Z"
     }
    },
-   "execution_count": 18
+   "execution_count": 9
   },
   {
    "cell_type": "markdown",
@@ -430,11 +430,11 @@
    "metadata": {
     "collapsed": false,
     "ExecuteTime": {
-     "end_time": "2024-06-03T13:57:39.109392Z",
-     "start_time": "2024-06-03T13:57:39.106462Z"
+     "end_time": "2024-06-17T13:07:22.231563Z",
+     "start_time": "2024-06-17T13:07:22.228483Z"
     }
    },
-   "execution_count": 18
+   "execution_count": 9
   }
  ],
  "metadata": {

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -105,6 +105,7 @@ dev = [
     "pre-commit",
     "pytest",
     "pytest-cov",
+    "pytest-mock",
     "pytest-randomly",
     "pytest-timeout",
     "pytest-xdist[psutil]",

